Community : Add audio-parser "faster-whisper" in audio.py

hulitaitai opened the pull request:
faster-whisper is a reimplementation of OpenAI's Whisper model using CTranslate2, which is up to 4 times faster than enai/whisper for the same accuracy while using less memory. The efficiency can be further improved with 8-bit quantization on both CPU and GPU.

It can automatically detect the following 14 languages and transcribe the text into their respective languages: en, zh, fr, de, ja, ko, ru, es, th, it, pt, vi, ar, tr.

The gitbub repository for faster-whisper is :
    https://github.com/SYSTRAN/faster-whisper

hulitaitai committed:
Update text2vec.py

Add that URL of the embedding tool "text2vec".
Fix minor mistakes in the doc-string.

hulitaitai committed:
Community : Add audio-parser "faster-whisper" in audio.py

faster-whisper is a reimplementation of OpenAI's Whisper model using CTranslate2, which is up to 4 times faster than enai/whisper for the same accuracy while using less memory. The efficiency can be further improved with 8-bit quantization on both CPU and GPU.

It can automatically detect the following 14 languages and transcribe the text into their respective languages: en, zh, fr, de, ja, ko, ru, es, th, it, pt, vi, ar, tr.

The gitbub repository for faster-whisper is :
    https://github.com/SYSTRAN/faster-whisper

hulitaitai committed:
Community : Add audio-parser "faster-whisper" in audio.py

faster-whisper is a reimplementation of OpenAI's Whisper model using CTranslate2, which is up to 4 times faster than enai/whisper for the same accuracy while using less memory. The efficiency can be further improved with 8-bit quantization on both CPU and GPU.

It can automatically detect the following 14 languages and transcribe the text into their respective languages: en, zh, fr, de, ja, ko, ru, es, th, it, pt, vi, ar, tr.

The gitbub repository for faster-whisper is :
https://github.com/SYSTRAN/faster-whisper

hulitaitai commented:
I am sorry, I have never passed the lint test successfully. I really don't know what is wrong.

faster-whisper is an open-source parser, and it is a low-cost parser which can run with less than 5GB of GPU memory.

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+
+    def __init__(
+        self,
+        device: Optional[str] = "0",
```
```suggestion
        *,
        device: Optional[str] = "0",
```

hulitaitai replied:
Yes, here, I presumed I would reassign the definite value later, so I casually placed a value here. 

Additionally, I admit I copied a lot of code and idees from the parsers above in this file. The code I've written runs well on my PC. I will meticulously re-examine various transmission and variables as you reminde, for the sake of overall system. I'll come back to you after I've finished checking.

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+
+    def __init__(
+        self,
+        device: Optional[str] = "0",
```
device "0" doesn't look like correct value given that it can only end up being cuda or cpu with logic below.

eyurtsev replied:
could you document the arguments via an init?

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+            )
+
+        # Determine the device to use
+        if device == "cpu":
```
logic looks like it's ignoring input from user

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+
+    The gitbub repository for faster-whisper is :
+    https://github.com/SYSTRAN/faster-whisper
+
```
If possible it'll be great to include a usage Example with the full import using a ..code-block: python

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+    def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+        """Lazily parse the blob."""
+
+        import io
```
can be in global namespace

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+            )
+
+        # Audio file from disk
+        audio = AudioSegment.from_file(blob.path)
```
Could you check that a .path is set on the blob? Blobs do not have to be associated with a file in the file system as they can live just in memory.


Is it possible to handle using `blob.as_bytes()`.

hulitaitai replied:
In the blob sent by ffmpeg, there is only the path of the audio file and nothing in its metadata.

Faster-whisper does not transcribe the bytes, it transcribes only mp3 files. So it should ensure that the file sent to it is an mp3 file.

eyurtsev replied:
The parsing is decoupled from the code that loads the blob. To be correct, the parser cannot assume how the blob was generated instead it should assume blob against the blob API (see link below) -- the blob may be associated with metadata and the blob may not exist on file but instead in memory.

Here's the API for the blob

https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/document_loaders/blob_loaders.py#L40-L40

Let me know if you have any questions!

So things that need to be done here:

1. Propagate metadata from the blob
2. Make sure that the parser keeps working if the blob was only specified as binary data (using the data attribute) rather than generated from an existing file

eyurtsev replied:
@hulitaitai if you're willing, I can make the needed changes, but I'll need your help to help test it

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+        for segment in segments:
+            yield Document(
+                page_content=segment.text,
+                metadata={"source": blob.source},
```
{
  'source': blob.source,
  **blob.metadata, # To use existing metadata
}

eyurtsev commented on libs/community/langchain_community/document_loaders/parsers/audio.py:
```diff
@@ -319,3 +319,96 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
+            import torch
+        except ImportError:
+            raise ImportError(
+                "torch package not found, please install it with " "`pip install torch`"
```
```suggestion
                "torch package not found, please install it with `pip install torch`"
```

eyurtsev commented:
i can help with the linter -- i mostly care about some of the other functional changes 

https://python.langchain.com/docs/contributing/code/#formatting-and-linting

`make format` and `make lint` will help

hulitaitai committed:
Add comments and examples, and review the usage of the 'blob'

In the blob sent by ffmpeg, there is only the path of the audio file and nothing in its metadata.

Faster-whisper transcribes only mp3 files, so it should ensure that the file sent to it is an mp3 file.

I also added timestamps in the metadata returned to other LangChain functions.

hulitaitai commented:
> i can help with the linter -- i mostly care about some of the other functional changes
> 
> https://python.langchain.com/docs/contributing/code/#formatting-and-linting
> 
> `make format` and `make lint` will help

I asked ChatGPT how to use 'make format' and 'make lint'. I will use it before my next Pull Request.

hulitaitai committed:
Ensure "self.model_size" is always correct

Ensure that the variable self.model_size is always assigned a correct value.

hulitaitai committed:
Add verification of the origin of the audio and complete the metadata

1. The audio can come from the data in the form of bytes, or from the path.
2. More information is included in the metadata, and the origin metadata is propagated.

hulitaitai committed:
Merge branch 'master' into master

hulitaitai commented:
I have tested the changes with the following example, it works.

from pathlib import Path
from langchain_community.document_loaders.blob_loaders import Blob
from langchain_community.document_loaders.parsers.audio import FasterWhisperParser

m4a_path = Path("example.m4a")

with open(m4a_path, "rb") as f:
    m4a_bytes = f.read()

# blob built with binary data
blob = Blob.from_data(data=m4a_bytes)

p = FasterWhisperParser()
docs = p.lazy_parse(blob)
for doc in docs:
    print(doc)

hulitaitai commented:
I have tested the changes with the following example, it works.

from pathlib import Path
from langchain_community.document_loaders.blob_loaders import Blob
from langchain_community.document_loaders.parsers.audio import FasterWhisperParser

m4a_path = Path("example.m4a")

with open(m4a_path, "rb") as f:
m4a_bytes = f.read()

blob = Blob.from_data(data=m4a_bytes)

p = FasterWhisperParser()
docs = p.lazy_parse(blob)
for doc in docs:
    print(doc)

hulitaitai commented:
I have tested the changes with with a blob built with "Blob.from_data(data=m4a_bytes)", and it works.

hulitaitai commented:
I have tested the changes with with a blob built with "Blob.from_data(data=m4a_bytes)", and the parser works.

eyurtsev committed:
Merge branch 'master' into master

eyurtsev committed:
x

eyurtsev commented:
Looks great! I'll resolve the linting issue tomorrow.

If you end up working with `blob` again, it has a convenience interface called `blob.as_bytes_io`

```
    @contextlib.contextmanager
    def as_bytes_io(self) -> Generator[Union[BytesIO, BufferedReader], None, None]:
        """Read data as a byte stream."""
        if isinstance(self.data, bytes):
            yield BytesIO(self.data)
        elif self.data is None and self.path:
            with open(str(self.path), "rb") as f:
                yield f
        else:
            raise NotImplementedError(f"Unable to convert blob {self}")
```

eyurtsev committed:
Merge branch 'master' into hulitaitai/master
